Cubical/HITs/Join/Properties.agda

```diff
@@ -323,3 +323,99 @@ joinSwitch = isoToEquiv (iso map map invol invol)
           ; (l = i1) → push (push a b i) c j
           })
         (push (push a b i) c j))
+
+{-
+  Direct proof of associativity.
+-}
+joinAssocDirect : ∀ {ℓ ℓ' ℓ''} {A : Type ℓ} {B : Type ℓ'} {C : Type ℓ''}
+  → join (join A B) C ≃ join A (join B C)
+joinAssocDirect {A = A} {B} {C} =
+  isoToEquiv (iso forward back forwardBack backForward)
+  where
+  forward : join (join A B) C → join A (join B C)
+  forward (inl (inl a)) = inl a
+  forward (inl (inr b)) = inr (inl b)
+  forward (inl (push a b i)) = push a (inl b) i
+  forward (inr c) = inr (inr c)
+  forward (push (inl a) c j) = push a (inr c) j
+  forward (push (inr b) c j) = inr (push b c j)
+  forward (push (push a b i) c j) =
+    hcomp
+      (λ k → λ
+        { (i = i0) → push a (inr c) (j ∧ k)
+        ; (i = i1) → inr (push b c j)
+        ; (j = i0) → push a (inl b) i
+        ; (j = i1) → push a (inr c) (i ∨ k)
+        })
+      (push a (push b c j) i)
+
+  back : join A (join B C) → join (join A B) C
+  back (inl a) = inl (inl a)
+  back (inr (inl b)) = inl (inr b)
+  back (inr (inr c)) = inr c
+  back (inr (push b c j)) = push (inr b) c j
+  back (push a (inl b) i) = inl (push a b i)
+  back (push a (inr c) i) = push (inl a) c i
+  back (push a (push b c j) i) =
+    hcomp
+      (λ k → λ
+        { (i = i0) → push (inl a) c (j ∧ ~ k)
+        ; (i = i1) → push (inr b) c j
+        ; (j = i0) → inl (push a b i)
+        ; (j = i1) → push (inl a) c (i ∨ ~ k)
+        })
+      (push (push a b i) c j)
+
+  forwardBack : ∀ u → forward (back u) ≡ u
+  forwardBack (inl a) = refl
+  forwardBack (inr (inl b)) = refl
+  forwardBack (inr (inr c)) = refl
+  forwardBack (inr (push b c j)) = refl
+  forwardBack (push a (inl b) i) = refl
+  forwardBack (push a (inr c) i) = refl
+  forwardBack (push a (push b c j) i) l =
+    comp
```

Comment:
Doesn't hcomp work here?

Reply:
Nope. These arise from applying forward/back to an hcomp, which produces a comp.

Reply:
Specifically, the `l = i0` face of the clause would not be correct if we used an hcomp.

Reply:
I see. In this case I guess it could have been an hcomp though, as forward/back are non-dependent functions...

Reply:
I mean that Cubical Agda could in principle have reduced to an hcomp instead of a comp when applying forward/back to an hcomp

Reply:
yep